README.md
# @CentralPing/passcode

[![Build Status](https://travis-ci.org/CentralPing/passcode.svg?branch=master)](https://travis-ci.org/CentralPing/passcode)
[![Coverage Status](https://coveralls.io/repos/github/CentralPing/passcode/badge.svg)](https://coveralls.io/github/CentralPing/passcode)
[![Dependency Status](https://david-dm.org/CentralPing/passcode.svg)](https://david-dm.org/CentralPing/passcode)
[![Greenkeeper Status](https://badges.greenkeeper.io/CentralPing/passcode.svg)](https://greenkeeper.io/)
[![Known Vulnerabilities](https://snyk.io/test/github/centralping/passcode/badge.svg)](https://snyk.io/test/github/centralping/passcode)

A slightly opinionated stateless passcode manager.

Why *slightly* opinionated? Some people like more flexibility to implement solutions and this module hopefully allows for that. The few opinions employed in this module include the choice of hashing method, [`pbkdf2`](https://nodejs.org/api/crypto.html#crypto_crypto_pbkdf2sync_password_salt_iterations_keylen_digest) (which can be configured), and a custom random 6 digit passcode generator (custom passcodes can also be provided). The biggest opinion however is to enforce signing the JWT and hashing the passcode. Doing niether results in an unsecure and effectively useless passcode verification as unsigned JWTs can be modified by anyone and unhashed codes can be read by anyone. If that is desired then this module is superflous as the underlying [jsonwebtoken](https://github.com/auth0/node-jsonwebtoken) module can handle that directly.

## Installation

`npm i --save @centralping/passcode`

## API Reference

<a name="module_passcode..issue"></a>

### passcode~issue(security, [payload], claims, hash) ⇒ <code>Object</code>
Issues a signed token with a hashed passcode, token ID, expiration time
 and the stringified passcode.

**Kind**: inner method of [<code>passcode</code>](#module_passcode)  
**Returns**: <code>Object</code> - `{error, value: {id, expires, passcode, token}}`  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| security | <code>Object</code> |  |  |
| security.salt | <code>String</code> |  | A string to salt the passcode hash. |
| security.secret | <code>String</code> \| <code>Object</code> |  | A string or key to sign the JWT. |
| [security.passcode] | <code>String</code> |  | The passcode (defaults to random  6 digit string). |
| [security.iss] | <code>String</code> |  | JWT issuer (used as additional verification). |
| [security.aud] | <code>String</code> |  | JWT audience (used as additional  verification). |
| [security.sub] | <code>String</code> |  | JWT subject (used as additional  verification). |
| [payload] | <code>Object</code> |  | Essentially a passthrough for [jsonwebtoken](https://github.com/auth0/node-jsonwebtoken#jwtsignpayload-secretorprivatekey-options-callback)  payload support. |
| [payload.jti] | <code>Date</code> |  | JWT ID (defaults to a [`uuid.v4`](https://github.com/kelektiv/node-uuid#version-4) value). |
| claims | <code>Object</code> |  | Essentially a passthrough for [jsonwebtoken](https://github.com/auth0/node-jsonwebtoken#jwtsignpayload-secretorprivatekey-options-callback)  claims support. |
| [claims.expiresIn] | <code>Number</code> \| <code>String</code> | <code>5m</code> | JWT expiration time span.  In seconds or a parsable string for [zeit/ms](https://github.com/zeit/ms) |
| hash | <code>Object</code> |  | Essentially a passthrough for [`pbkdf2`](https://nodejs.org/api/crypto.html#crypto_crypto_pbkdf2sync_password_salt_iterations_keylen_digest)  hashing options. |
| [hash.iterations] | <code>Number</code> | <code>1000</code> | Hash iterations. |
| [hash.keyLength] | <code>Number</code> | <code>64</code> | Hash length. |
| [hash.digest] | <code>String</code> | <code>sha512</code> | Hashing algorithm. |
| [hash.encoding] | <code>String</code> | <code>hex</code> | Hash encoding. |

**Example**  
```js
const {error, value} = issue({salt, secret}, {email: 'foo@bar.com'});
```
<a name="module_passcode..verify"></a>

### passcode~verify(token, security, hash) ⇒ <code>Object</code>
Verifies a signed token with the provided challenge passcode.

**Kind**: inner method of [<code>passcode</code>](#module_passcode)  
**Returns**: <code>Object</code> - `{error, value: {iat, jti, exp, ...TOKEN_PAYLOAD}}`  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| token | <code>String</code> |  | The token to be verified |
| security | <code>Object</code> |  |  |
| security.passcode | <code>String</code> |  | The challenge passcode. |
| security.salt | <code>String</code> |  | A string to salt the challenge passcode hash. |
| security.secret | <code>String</code> \| <code>Object</code> |  | A string or key to verify the JWT. |
| [security.iss] | <code>String</code> |  | JWT issuer (used as additional verification). |
| [security.aud] | <code>String</code> |  | JWT audience (used as additional  verification). |
| [security.sub] | <code>String</code> |  | JWT subject (used as additional  verification). |
| hash | <code>Object</code> |  | Essentially a passthrough for [`pbkdf2`](https://nodejs.org/api/crypto.html#crypto_crypto_pbkdf2sync_password_salt_iterations_keylen_digest)  hashing options. |
| [hash.iterations] | <code>Number</code> | <code>1000</code> | Hash iterations. |
| [hash.keyLength] | <code>Number</code> | <code>64</code> | Hash length. |
| [hash.digest] | <code>String</code> | <code>sha512</code> | Hashing algorithm. |
| [hash.encoding] | <code>String</code> | <code>hex</code> | Hash encoding. |

**Example**  
```js
const {error, value} = verify(token, {passcode, salt, secret});
```

## Examples

### For Simple Verification With a Secret

```js
const {issue, verify} = require('passcode');

// Generate a signed token with hashed random passcode
const {error, value: {id, expires, passcode, token}} = issue({
  salt: YOUR_SALT,
  secret: YOUR_SECRET
});
/**
 * Do something with the token
 */
// Verify token with code
const {error, value} = verify(token, {
  passcode: ENTERED_PASSCODE,
  salt: YOUR_SALT,
  secret: YOUR_SECRET
});
```

### For Simple Verification With a Key

```js
const {issue, verify} = require('passcode');
const secret = fs.readFileSync('public.pem');

// Generate a signed token with hashed random passcode
const {error, value: {id, expires, passcode, token}} = issue({
  salt: YOUR_SALT,
  secret
});
/**
 * Do something with the token
 */
// Verify token with code
const {error, value} = verify(token, {
  passcode: ENTERED_PASSCODE,
  salt: YOUR_SALT,
  secret
});
```

### For Including Payload Data

```js
const {issue, verify} = require('passcode');

// Generate a signed token with custom payload
const {error, value: {id, expires, passcode, token}} = issue({
  salt: YOUR_SALT,
  secret: YOUR_SECRET
}, {
  email: 'foo@bar.com'
});
/**
 * Do something with the token
 */
// Verify token with code
const {error, value: {email}} = verify(token, {
  passcode: ENTERED_PASSCODE,
  salt: YOUR_SALT,
  secret: YOUR_SECRET
});
```

## License

MIT


CHANGELOG.md
#### v0.1.1 (2018-06-14)

##### Chores

* **package:**  include public publishing option for builds ([4c8dd0b1](https://github.com/CentralPing/passcode/commit/4c8dd0b1e5e2568e2b2a7af259efd7c842624087))

#### v0.1.0 (2018-06-14)

##### Chores

* **package:**
  *  namespace package ([fa3d54a6](https://github.com/CentralPing/passcode/commit/fa3d54a62f922e9fecdb801afe3c8dd0f3196f9f))
  *  update scripts for npm publishing ([0c4c72d3](https://github.com/CentralPing/passcode/commit/0c4c72d39eb58e6688d6729de6495773c9e911a1))
  *  rename integration test file pattern ([bbca601e](https://github.com/CentralPing/passcode/commit/bbca601e32dc75cdf45fef1e315c1e2c2c6a80b4))
  *  dependency bumps ([a7cdaff7](https://github.com/CentralPing/passcode/commit/a7cdaff78e8e8529188a3742fae8af84bc91faf7))
  *  update dependencies ([6ede95ea](https://github.com/CentralPing/passcode/commit/6ede95eabf2f6694bd11679f43bd411d19a6a589))
  *  update dependencies ([1985a66e](https://github.com/CentralPing/passcode/commit/1985a66e93ff88143015debf4258c9ffb0a3b071))
* **travis:**
  *  update build to publish on tagged releases ([0e2e2ae3](https://github.com/CentralPing/passcode/commit/0e2e2ae3481963b252a92d990e52de5f9b9da2e1))
  *  add travis integration ([cc674f7f](https://github.com/CentralPing/passcode/commit/cc674f7f55e2a67f6eec7ba4aed0073772d9d958))
* **npm:**  flag dev artifacts from inclusion in published package ([5083c03c](https://github.com/CentralPing/passcode/commit/5083c03c44534c006cfb212b083f6079924b7c20))
* **index:**  import correct module ([74604c15](https://github.com/CentralPing/passcode/commit/74604c153b1c75273a1a138d1dac9bb09348604b))
* **token:**  standardize naming ([9fa27f7f](https://github.com/CentralPing/passcode/commit/9fa27f7f1341381d201d79f7443c26e58ac5a7db))
*  fix script for lint; add coverage publishing ([8b2b107b](https://github.com/CentralPing/passcode/commit/8b2b107b4a6b5a952be4c103148ef7a2b9803d5c))
*  add coverage reporting ([fbdc09ce](https://github.com/CentralPing/passcode/commit/fbdc09cee1a191bce2e4acff6b42c42111c20538))

##### Documentation Changes

* **readme:**
  *  update docs for published npm install ([ec112261](https://github.com/CentralPing/passcode/commit/ec112261175309a04308b66852c25ed02c3146cd))
  *  add example for using a key as the secret ([2acaebfb](https://github.com/CentralPing/passcode/commit/2acaebfbf05eff2be0f42857b5c35feaa8ef8ebc))
  *  update docs for current usage ([a251c4b4](https://github.com/CentralPing/passcode/commit/a251c4b4082b3737783fa0c884b488a7c02891a4))
  *  update docs for current usage ([f917cb6f](https://github.com/CentralPing/passcode/commit/f917cb6f52d3b7f8586072bac14b66942dbd8ca5))
  *  update docs for current usage ([20a8b9d4](https://github.com/CentralPing/passcode/commit/20a8b9d4dd56be832cbd49482fbb6cf650ff3358))
  *  fix module name ([fccb622a](https://github.com/CentralPing/passcode/commit/fccb622acc36113d606f034e898f40368abffedf))
* **main, readme:**  clarify secret can also be a key; clean up description ([fceaf31d](https://github.com/CentralPing/passcode/commit/fceaf31da36225807d757124c339dccaa250dcf9))
* **main:**
  *  make esline compliant ([1d2cb3a3](https://github.com/CentralPing/passcode/commit/1d2cb3a33512f468904ac1299c4415f108ee9923))
  *  define parameters ([2c17ae71](https://github.com/CentralPing/passcode/commit/2c17ae719c431e62e921d7929b82a36080620767))
*  remove incorrect badge ([87b4703e](https://github.com/CentralPing/passcode/commit/87b4703ecd8b3836c002a7459cebdf12291cebc5))

##### New Features

* **main:**
  *  allow secrect to be optional ([c85baa3b](https://github.com/CentralPing/passcode/commit/c85baa3b7b3de2b1b9cd10cbb9e19f9c5fd6585b))
  *  removed superflous key in response object ([f1b6d9ba](https://github.com/CentralPing/passcode/commit/f1b6d9ba356a3c6b973bbf7427b61a177dc5a071))
  *  make code hashing optional ([94eab169](https://github.com/CentralPing/passcode/commit/94eab1691da7c99513044e598c9f2abdb5324ac6))
  *  allow custom codes for verification ([baa25f2c](https://github.com/CentralPing/passcode/commit/baa25f2c7ad1c5f6bdef92ebd08faad62de211ac))
* **token:**
  *  return error/value as object ([aab9fa17](https://github.com/CentralPing/passcode/commit/aab9fa17f470a9f681d0428981e4542fbe862196))
  *  remove config export ([7980ba04](https://github.com/CentralPing/passcode/commit/7980ba049583d7060cdab43a74b69703c6b67e40))
* **lib:**  export directly from token ([08e430dc](https://github.com/CentralPing/passcode/commit/08e430dc244554156f03bb25c2846e071d600c7e))
*  initial commit ([19a50526](https://github.com/CentralPing/passcode/commit/19a50526d6997e9ecaebd7152fcccf6ec283b75c))

##### Other Changes

* **token:**  make salt and secret required params ([e1632e4a](https://github.com/CentralPing/passcode/commit/e1632e4a9d531de29f773bbdfb4a04afbed272d6))

##### Refactors

* **main:**
  *  make passcode payload property a global ([1518f182](https://github.com/CentralPing/passcode/commit/1518f1822d151758feb050dba8eddc133d55ee7f))
  *  clean up parameters ([23959cc4](https://github.com/CentralPing/passcode/commit/23959cc4ac0940e94873ddd83ef75e40e6af0954))
  *  remove code from response object via destructuring ([5ea1e646](https://github.com/CentralPing/passcode/commit/5ea1e64671819d984caaaab6958160cd6f06c35e))

##### Tests

* **main:**
  *  make eslint compliant ([6d8c66e5](https://github.com/CentralPing/passcode/commit/6d8c66e57c38c4e28aea065b08877abb77aff5d9))
  *  make salt, secret and passcode are required ([4e9be154](https://github.com/CentralPing/passcode/commit/4e9be15424b921b4f386fd6566039f8a2cb05e23))
  *  ensure salt, secret and passcode are required ([7147d41f](https://github.com/CentralPing/passcode/commit/7147d41f3ed09d07e48a7bf1c1dc663ad2d08519))
  *  make eslint compliant ([e49bf094](https://github.com/CentralPing/passcode/commit/e49bf09430c0e3247735e550e5d1cb72a1203fcd))
  *  ensure secret is optional ([5f923a17](https://github.com/CentralPing/passcode/commit/5f923a17a2bb7fe92359952e048f88630d1a1256))

lib/main.js
/**
 * @module passcode
*/

const {sign: jwtSign, verify: jwtVerify} = require('jsonwebtoken');
const {pbkdf2Sync} = require('crypto');
const {v4} = require('uuid');

const {randomCode} = require('./random-code');
const {filter: filterObject} = require('./filter');

const CODE_PROP = '$CPPC';
const ITERATIONS = 1000;
const KEY_LENGTH = 64;
const DIGEST = 'sha512';
const ENCODING = 'hex';
const CODE_LENGTH = 6;
const EXPIRES_IN = '5m';

exports.issue = issue;
exports.verify = verify;

/**
 * Issues a signed token with a hashed passcode, token ID, expiration time
 *  and the stringified passcode.
 * @example
 * const {error, value} = issue({salt, secret}, {email: 'foo@bar.com'});
 * @param {Object} security
 * @param {String} security.salt A string to salt the passcode hash.
 * @param {String|Object} security.secret A string or key to sign the JWT.
 * @param {String} [security.passcode] The passcode (defaults to random
 *  6 digit string).
 * @param {String} [security.iss] JWT issuer (used as additional verification).
 * @param {String} [security.aud] JWT audience (used as additional
 *  verification).
 * @param {String} [security.sub] JWT subject (used as additional
 *  verification).
 * @param {Object} [payload] Essentially a passthrough for [jsonwebtoken](https://github.com/auth0/node-jsonwebtoken#jwtsignpayload-secretorprivatekey-options-callback)
 *  payload support.
 * @param {Date} [payload.jti] JWT ID (defaults to a [`uuid.v4`](https://github.com/kelektiv/node-uuid#version-4) value).
 * @param {Object} claims Essentially a passthrough for [jsonwebtoken](https://github.com/auth0/node-jsonwebtoken#jwtsignpayload-secretorprivatekey-options-callback)
 *  claims support.
 * @param {(Number|String)} [claims.expiresIn=5m] JWT expiration time span.
 *  In seconds or a parsable string for [zeit/ms](https://github.com/zeit/ms)
 * @param {Object} hash Essentially a passthrough for [`pbkdf2`](https://nodejs.org/api/crypto.html#crypto_crypto_pbkdf2sync_password_salt_iterations_keylen_digest)
 *  hashing options.
 * @param {Number} [hash.iterations=1000] Hash iterations.
 * @param {Number} [hash.keyLength=64] Hash length.
 * @param {String} [hash.digest=sha512] Hashing algorithm.
 * @param {String} [hash.encoding=hex] Hash encoding.
 * @return {Object} `{error, value: {id, expires, passcode, token}}`
 */
function issue(
  {
    passcode = randomCode(CODE_LENGTH),
    salt = new TypeError('"salt" is required'),
    secret = new TypeError('"secret" is required'),
    iss,
    aud,
    sub
  } = {},
  {
    jti = v4(),
    ...payload
  } = {},
  {
    expiresIn = EXPIRES_IN,
    ...claims
  } = {},
  {
    iterations = ITERATIONS,
    keyLength = KEY_LENGTH,
    digest = DIGEST,
    encoding = ENCODING
  } = {}
) {
  try {
    if (salt instanceof Error) throw salt;
    if (secret instanceof Error) throw secret;

    const challenge = pbkdf2Sync(
      passcode,
      salt,
      iterations,
      keyLength,
      digest
    ).toString(encoding);
    const tokenPayload = filterObject({
      iss,
      aud,
      sub,
      ...payload,
      jti,
      [CODE_PROP]: challenge
    }, ([key, value]) => value !== undefined);
    const tokenClaims = filterObject({
      ...claims,
      expiresIn,
      mutatePayload: true
    }, ([key, value]) => value !== undefined);
    const token = jwtSign(tokenPayload, secret, tokenClaims);

    return {
      error: null,
      value: {
        id: jti,
        expires: tokenPayload.exp,
        token,
        passcode
      }
    };
  } catch (error) {
    return {error};
  }
}

/**
 * Verifies a signed token with the provided challenge passcode.
 * @example
 * const {error, value} = verify(token, {passcode, salt, secret});
 * @param {String} token The token to be verified
 * @param {Object} security
 * @param {String} security.passcode The challenge passcode.
 * @param {String} security.salt A string to salt the challenge passcode hash.
 * @param {String|Object} security.secret A string or key to verify the JWT.
 * @param {String} [security.iss] JWT issuer (used as additional verification).
 * @param {String} [security.aud] JWT audience (used as additional
 *  verification).
 * @param {String} [security.sub] JWT subject (used as additional
 *  verification).
 * @param {Object} hash Essentially a passthrough for [`pbkdf2`](https://nodejs.org/api/crypto.html#crypto_crypto_pbkdf2sync_password_salt_iterations_keylen_digest)
 *  hashing options.
 * @param {Number} [hash.iterations=1000] Hash iterations.
 * @param {Number} [hash.keyLength=64] Hash length.
 * @param {String} [hash.digest=sha512] Hashing algorithm.
 * @param {String} [hash.encoding=hex] Hash encoding.
 * @return {Object} `{error, value: {iat, jti, exp, ...TOKEN_PAYLOAD}}`
 */
function verify(
  token = new TypeError('"token" is required'),
  {
    passcode = new TypeError('"passcode" is required'),
    salt = new TypeError('"salt" is required'),
    secret = new TypeError('"secret" is required'),
    iss: issuer,
    aud: audience,
    sub: subject
  } = {},
  {
    iterations = ITERATIONS,
    keyLength = KEY_LENGTH,
    digest = DIGEST,
    encoding = ENCODING
  } = {}
) {
  try {
    if (token instanceof Error) throw token;
    if (passcode instanceof Error) throw passcode;
    if (salt instanceof Error) throw salt;
    if (secret instanceof Error) throw secret;

    const {[CODE_PROP]: $CPPC, ...decoded} = jwtVerify(
      token,
      secret,
      {
        issuer,
        audience,
        subject
      }
    );
    const challenge = pbkdf2Sync(
      passcode,
      salt,
      iterations,
      keyLength,
      digest
    ).toString(encoding);

    if ($CPPC !== challenge) {
      throw new Error('Invalid challenge passcode');
    }

    return {error: null, value: decoded};
  } catch (error) {
    return {error};
  }
}
